src/node/audio_buffer_source.rs

```diff
@@ -0,0 +1,673 @@
+                default_value: 1.,
+                automation_rate: AutomationRate::K,
+            };
+            let (pr_param, pr_proc) = context
+                .base()
+                .create_audio_param(playback_rate_param_options, registration.id());
+
+            pr_param.set_value(playback_rate);
+
+            // Channel to send buffer channels references to the renderer
+            //
+            // @note: Maybe we don't want to block the control thread waiting for next
+            // `render.tick`? However it's probably better to block the control thread
+            // rather than the render thread...
+            // Wouldn't be feasible to spawn a middle man thread that could block?
+            // ```
+            // let (sender, receiver) = crossbeam_channel::bounded(0);
+            // thread::spawn(move || {
+            //     sender.send(data); // this thread could block waiting for the render thread
+            // });
+            // ```
+            // ...but maybe this could introduce strange race conditions, or
+            // performance penalty, I don't know...
+            //
+            // @note - for some reason it blocks the rendering in tests w/ offline audio context
+            //  maybe because of early return in case of `none` buffer
+            let (sender, receiver) = crossbeam_channel::bounded(1);
+
+            let shared_attributes = SharedAttributes::default();
+
+            let renderer = AudioBufferSourceRenderer {
+                attributes: shared_attributes.clone(),
+                receiver,
+                // @todo - use an `AudioBuffer` directly
+                // need to make sure how `clone` behave with nested `Arc` structures
+                buffer: None,
+                buffer_duration: None,
+                detune: d_proc,
+                playback_rate: pr_proc,
+                render_state: AudioBufferRendererState::default(),
+                // 2 channels is sufficient for most use-cases, the `vec` will be
+                // resized to actual buffer number_of_channels when buffer is received
+                // on the render thread
+                internal_buffer: Vec::<f32>::with_capacity(2),
+            };
+
+            let mut node = Self {
+                registration,
+                attributes: shared_attributes,
+                channel_config: channel_config.into(),
+                sender,
+                detune: d_param,
+                playback_rate: pr_param,
+                buffer: None,
+                source_started: false,
+            };
+
+            node.attributes.set_loop(loop_);
+            node.attributes.set_loop_start(loop_start);
+            node.attributes.set_loop_end(loop_end);
+
+            if let Some(buf) = buffer {
+                node.set_buffer(&buf);
+            }
+
+            (node, Box::new(renderer))
+        })
+    }
+
+    pub fn set_buffer(&mut self, audio_buffer: &AudioBuffer) {
```

Comment:
I dont think the spec allows setting buffer multiple times

Reply:
the spec allows to do 
```js
const src = audioContext.createBufferSource()
src.buffer = myBuffer;
```

Reply:
huh, somehow I missed the code below where you guard for setting it multiple times. Never mind this